Introduce runtime section in mappings

javanna opened the pull request:
The runtime section is at the same level as the existing properties section. Its purpose is to hold runtime fields only. With the introduction of the runtime section, a runtime field can be defined by specifying its type (previously called runtime_type) and script.

```
PUT /my-index/_mappings
{
    "runtime" : {
        "day_of_week" : {
            "type" : "keyword",
            "script" : {
                "source" : "emit(doc['timestamp'].value.dayOfWeekEnum.getDisplayName(TextStyle.FULL, Locale.ROOT))"
            }
        }
    },
    "properties" : {
        "timestamp" : {
            "type" : "date"
        }
    }
}
```

Fields defined in the runtime section can be updated at any time as they are not present in the lucene index. They get replaced entirely when they get updated.

Thanks to the introduction of the runtime section, runtime fields override existing mapped fields defined with the same name, similarly to runtime fields defined in the search request.


Note that the get field mappings API does not return runtime fields as it iterates through the mappers directly, which have no visibility over runtime fields.

Relates to #59332

nik9000 reviewed (commented):
:+1: If we want to go this way I think this is how we should do it.

nik9000 commented:
> +1 If we want to go this way I think this is how we should do it.

And by that I mean, I don't see any problems with the code. I'm not 100% sure we want the new section, but if we want it then this implements it how I expect.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -51,26 +51,36 @@ private static void put(Map<String, Analyzer> analyzers, String key, Analyzer va
+        collect(mapping.root, newObjectMappers, newFieldMappers, newFieldAliasMappers);
+
+        //add runtime fields and replace concrete fields with runtime fields with the same name
+        for (RuntimeFieldMapper runtimeMapper : mapping.root.runtimeMappers()) {
```
How does this work when parsing documents? We presumably want to ensure that an incoming doc is parsed using the non-runtime version of the mapper if this exists?

javanna replied:
good point. I thought that I tested this, but looking at the code I think replacing the field mappers should also affect documents parsing which I think is not what we want.

romseygeek replied:
Yeah, I think we need to do this via MappedFieldType, not via FieldMapper, unfortunately.  Which means we need to move `valueFetcher()` - I can make a start on that today.

javanna replied:
I am not sure that valueFetcher is the only problem, I would be happy if that was the case.

So the idea could be to only replace field mappers that are provided to `FieldTypeLookup`? And go through the usages of `MappingLookup#getMapper` and `MappingLookup#iterator` to see if they may need to access the runtime fields or not?

romseygeek replied:
Yes, that sounds reasonable.  Or we could even pass the runtime mappers directly to FieldTypeLookup, like we do currently with aliases, and re-map things in there?

javanna replied:
yea that would work too, ok I will try to find out what else may be on our way here.

javanna committed:
Add support for runtime section in the mappings

The runtime section is at the same level as the existing properties section. Its purpose is to hold runtime fields only. With the introduction of the runtime section, a runtime field can be defined by specifying its type (previously called runtime_type) and script.

Fields defined in the runtime section can be updated at any time as they are not present in the lucene index. They get replaced entirely when they get updated.

Thanks to the introduction of the runtime section, runtime fields can override existing mapped fields defined with the same name, similarly to runtime fields defined in the search request.

javanna committed:
iter

javanna committed:
fix PermissionsIT

javanna committed:
Merge branch 'master' into enhancement/mapping_runtime_section

javanna committed:
Merge branch 'master' into enhancement/mapping_runtime_section

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/RuntimeFieldType.java:
```diff
@@ -0,0 +1,96 @@
+            throws MapperParsingException;
+    }
+
+    public static void parseRuntimeFields(Map<String, Object> node,
```
I am not a big fan of this public static method here, but I think it's important to share it between QueryShardContext and RootObjectMapper

javanna commented on server/src/test/java/org/elasticsearch/index/IndexSortSettingsTests.java:
```diff
@@ -129,4 +144,50 @@ public void testInvalidMissing() {
             " must be one of [_last, _first]"));
     }
+
+    public void testIndexSorting() {
```
this was moved from RuntimeFieldMapperTests

javanna commented on x-pack/plugin/runtime-fields/src/test/java/org/elasticsearch/xpack/runtimefields/mapper/AbstractScriptFieldTypeTestCase.java:
```diff
@@ -6,29 +6,173 @@
+        );
+    }
+
+    public final void testMinimalMappingToMaximal() throws IOException {
```
these three tests I borrowed from MapperTestCase , they were previously run as part of RuntimeFieldMapperTests

javanna commented on x-pack/plugin/runtime-fields/src/test/java/org/elasticsearch/xpack/runtimefields/mapper/AbstractScriptFieldTypeTestCase.java:
```diff
@@ -6,29 +6,173 @@
+        assertEquals("Failed to parse mapping: stored scripts are not supported for runtime field [field]", exception.getMessage());
+    }
+
+    public void testFieldCaps() throws Exception {
```
these were moved from RuntimeFieldMapperTests

javanna commented on x-pack/plugin/runtime-fields/src/test/java/org/elasticsearch/xpack/runtimefields/mapper/DateScriptFieldTypeTests.java:
```diff
@@ -63,8 +67,43 @@
+        MappedFieldType fieldType = mapperService.fieldType("field");
+        assertThat(fieldType, instanceOf(DateScriptFieldType.class));
+        assertEquals(Strings.toString(mapping.get()), Strings.toString(mapperService.documentMapper()));
+    }
```
these were moved from RuntimeFieldMapperTests

javanna committed:
iter

javanna committed:
import

romseygeek commented on x-pack/plugin/runtime-fields/src/test/java/org/elasticsearch/xpack/runtimefields/mapper/AbstractScriptFieldTypeTestCase.java:
```diff
@@ -6,29 +6,170 @@
 import static org.mockito.Mockito.when;
 
-abstract class AbstractScriptFieldTypeTestCase extends ESTestCase {
+public abstract class AbstractScriptFieldTypeTestCase extends ESTestCase {
```
Dumb question: is there a reason this can't extend MappingServiceTestCase?

javanna replied:
A question I asked myself too. I think that MapperServiceTestCase revolves around ordinary fields that have a field mapper and a field type. Runtime fields are created differently and don't have a field mapper. I did not feel like extending MapperServiceTestCase to support runtime fields. I can see how we may want to reconsider this once we have more runtime field types

romseygeek replied:
`MapperTestCase` expects a FieldMapper, but `MapperServiceTestCase` just provides the helper functions and doesn't have any abstract methods on it.  Would save copying the `createMapperService()` methods.

javanna replied:
that works! I had tried MapperTestCase because I was annoyed that I had to copy methods from it, but with MapperServiceTestcase at least I don't have to copy createMapperService.

javanna committed:
fix tests

nik9000 reviewed (commented):
I left some formatting stuff  but nothing big at all. Let me know if you'd like a consult on the sneaky tests.

javanna committed:
iter

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java:
```diff
@@ -19,25 +19,68 @@
+            indicesModule.getMapperRegistry(), () -> {
+                throw new UnsupportedOperationException();
+            }, () -> true, null);
+    }
```
yes, I had to rework stuff a bit but it works.

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java:
```diff
@@ -465,13 +503,216 @@ public void testIllegalDynamicTemplate7DotXIndex() throws Exception {
+    public void testRuntimeSection() throws IOException {
+        MapperService mapperService = createMapperService(Collections.singletonList(new RuntimeFieldPlugin()));
+        MergeReason reason = randomFrom(MergeReason.MAPPING_UPDATE, MergeReason.INDEX_TEMPLATE);
+        String mapping = Strings.toString(XContentFactory.jsonBuilder()
```
on it

javanna commented on x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/RuntimeFields.java:
```diff
@@ -6,29 +6,55 @@
-        return Collections.singletonMap(RuntimeFieldMapper.CONTENT_TYPE, RuntimeFieldMapper.PARSER);
+    public Map<String, RuntimeFieldType.Parser> getRuntimeFieldTypes() {
+        return Map.of(
+            BooleanFieldMapper.CONTENT_TYPE,
```
it seems overkill to do that so that the formatter does the right thing?

nik9000 replied:
It makes the code a ton easier to read which I like.

javanna replied:
done

javanna commented on x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/AbstractScriptFieldType.java:
```diff
@@ -12,52 +12,58 @@
-    public final String familyTypeName() {
-        return runtimeType();
+        this.script = script;
+        this.toXContent = toXContent;
```
so, this is slightly hacky, it is done this way so that we get the Parameter serialization for free. We reuse the FieldMapper parsing code without having a FieldMapper. The constructors that take a function are there only for testing, because I don't want to have to create a FieldMapper.Builder in our tests.

javanna commented on x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/AbstractScriptFieldType.java:
```diff
@@ -174,46 +180,115 @@ public SpanQuery spanPrefixQuery(String value, SpanMultiTermQueryWrapper.SpanRew
+     *  {@link RuntimeFieldTypeParser#parse(String, Map, Mapper.TypeParser.ParserContext)} calls
+     *  {@link Builder#parse(String, Mapper.TypeParser.ParserContext, Map)} and returns the corresponding
+     *  {@link org.elasticsearch.index.mapper.MappedFieldType}.
      */
```
ehm, doing what?

nik9000 replied:
Using the Builder at all. It feels like it makes sense because we have them now, but maybe its overkill?

javanna replied:
I like using Parameters to define what gets parsed and to reuse that part of the existing mappings infra. I think that not using them would mean going back to manual parsing and printing.

nik9000 replied:
Me too! I just wonder if we can avoid building the FieldMapper then. Like, in a follow up? Or not. I dunno. It just fields weird to build a dummy field mapper to make parsing easier. No big deal.

javanna replied:
we do not build the field mapper at all!

javanna replied:
pressed enter too quickly: we skip calling build on the builder. we make the builder create the right field type and return that, without creating the field mapper.

javanna replied:
one follow-up we could thin of to clean things up is to extract the parsing bits from FieldMapper.Builder as well as the Parameter class so that parsing can be done outside of a Builder.

javanna commented on x-pack/plugin/src/test/resources/rest-api-spec/test/runtime_fields/10_keyword.yml:
```diff
@@ -8,49 +8,46 @@ setup:
                     }
                   params:
                     prefix: node_
+            properties:
```
I actually never considered runtime fields could not be on top :) I would expect them to appear first because they can override ordinary fields, I guess.

javanna committed:
iter

javanna committed:
Merge branch 'master' into enhancement/mapping_runtime_section

javanna committed:
spotless

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

javanna commented:
I made this no longer a draft, it is ready for proper review. I listed some of the followups needed in the description above. The first one will be to support defining runtime fields from a dynamic template, which will also allow us to re-enable the qa tests for runtime fields defined in the mappings.

javanna committed:
formatting

javanna committed:
fix test

javanna committed:
Merge branch 'master' into enhancement/mapping_runtime_section

javanna committed:
Merge branch 'master' into enhancement/mapping_runtime_section

javanna committed:
fix test

romseygeek reviewed (approved):
A couple of queries, but LGTM as a whole.

javanna committed:
expand tests

javanna committed:
Merge branch 'master' into enhancement/mapping_runtime_section

javanna committed:
add comment

javanna committed:
fix test

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/FieldTypeLookupTests.java:
```diff
@@ -101,28 +104,94 @@ public void testSourcePathsWithCopyTo() {
+        {
+            Set<String> sourcePaths = fieldTypeLookup.sourcePaths("field2");
+            assertEquals(1, sourcePaths.size());
+            assertTrue(sourcePaths.contains("field2"));
```
I see what you mean, though the logic in the method is very simple and I am reluctant to special case runtime fields, under the assumption that they are never in source. Also runtime fields can be plugged in and maybe we should not make assumptions about how they will fetch their values?

nik9000 replied:
I figure there is a chance runtime fields will use the source path, yeah. But at least for now they don't so it doesn't really matter. I *guess* emptySet is more correct, but, yeah, it doesn't really matter.

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java:
```diff
@@ -465,13 +447,127 @@ public void testIllegalDynamicTemplate7DotXIndex() throws Exception {
+            RuntimeField field2 = (RuntimeField)mapperService.fieldType("field2");
+            assertNull(field2.prop1);
+            assertNull(field2.prop2);
+        }
```
done

javanna commented on x-pack/plugin/runtime-fields/qa/core-with-search/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/test/search/CoreTestsWithSearchRuntimeFieldsIT.java:
```diff
@@ -116,20 +111,22 @@ protected boolean modifySearch(ApiCallSection search) {
                         return mergeMappings(new String[] { "*" });
                     }
                     String[] patterns = Arrays.stream(index.split(",")).map(m -> m.equals("_all") ? "*" : m).toArray(String[]::new);
+                    // TODO this is always false?
```
@nik9000 this is for you I think :)

nik9000 replied:
ooops. I think I mean `length == 1`. I can take a look in a follow up.

nik9000 commented on server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java:
```diff
@@ -150,51 +153,50 @@ public void testDynamicTemplates() throws Exception {
-                        .endObject()
-                    .endObject()
-                .endObject()
+            .startObject()
```
This one got harder to read!

javanna replied:
will fix

nik9000 commented on x-pack/plugin/runtime-fields/qa/build.gradle:
```diff
@@ -1,4 +1,4 @@
-// Shared infratructure
+// Shared infrastructure
```
Thanks!

nik9000 commented on x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/RuntimeFields.java:
```diff
@@ -6,29 +6,48 @@
-    public Map<String, Mapper.TypeParser> getMappers() {
-        return Collections.singletonMap(RuntimeFieldMapper.CONTENT_TYPE, RuntimeFieldMapper.PARSER);
+    public Map<String, RuntimeFieldType.Parser> getRuntimeFieldTypes() {
+        return Map.ofEntries(
```
:heart:

javanna committed:
restore formatting
